src/coreclr/jit/gentree.cpp

```diff
@@ -10213,6 +10213,72 @@ void Compiler::gtUpdateNodeSideEffects(GenTree* tree)
     });
 }
 
+//------------------------------------------------------------------------
+// gtSubTreeAndChildrenAreFirstExecutedEffects: Check whether the subtree is the first
+//                                              executed effect in a tree
+//
+// Arguments:
+//    tree - Tree to check the side effects on
+//    subTree - The subtree to be the first executed effect
+//
+// Returns:
+//    A boolean that indicates whether the subtree is the first executed effect in a tree
+//
+bool Compiler::gtSubTreeAndChildrenAreFirstExecutedEffects(GenTree* tree, GenTree* subTree)
+{
+    struct Visitor : GenTreeVisitor<Visitor>
+    {
+        enum
+        {
+            DoPreOrder        = true,
+            DoPostOrder       = true,
+            UseExecutionOrder = true,
+        };
+
+        Visitor(Compiler* comp, GenTree* subTree)
+            : GenTreeVisitor(comp)
+            , m_subTree(subTree)
+        {
+        }
+
+        fgWalkResult PreOrderVisit(GenTree** use, GenTree* user)
+        {
+            if (*use == m_subTree)
+            {
+                Result = true;
+                return WALK_ABORT;
+            }
+
+            return WALK_CONTINUE;
+        }
+
+        fgWalkResult PostOrderVisit(GenTree** use, GenTree* user)
+        {
+            if (((*use)->gtFlags & GTF_ALL_EFFECT) != GTF_EMPTY)
+            {
+                Result = false;
+                return WALK_ABORT;
+            }
+
+            if ((*use)->OperIsAnyLocal() && m_compiler->lvaGetDesc((*use)->AsLclVarCommon())->lvHasLdAddrOp)
```

Comment:
This check is specific to your case where it is called early in the phase pipeline. If we want this to be a general utility we might want to make it optional. At least we should leave a comment here.

Reply:
Added an optional parameter for this.

Reply:
`fgHeadTailMerge` just calls it `early`

Reply:
Done.